Add tests for static async functions in traits

bryangarza opened the pull request:
This patch adds test cases for AFIT, the majority of which are currently expected to run as `check-fail`.

---

Note: I grabbed the cases from https://hackmd.io/SwRcXCiWQV-WRJ4BYs53fA

Also, I'm not sure if the `async-associated-types2` and `async-associated-types2-desugared` are correct, I modified them a bit from the examples in the HackMD.

bryangarza committed:
Add tests for static async functions in traits

This patch adds test cases for AFIT, the majority of which are currently
expected to run as `check-fail`.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

bryangarza commented:
r? @compiler-errors

compiler-errors commented on src/test/ui/async-await/in-trait/async-associated-types.rs:
```diff
@@ -0,0 +1,24 @@
+// edition: 2021
```
This should have a bug filed for it I think

bryangarza replied:
Filed https://github.com/rust-lang/rust/issues/102681

compiler-errors commented on src/test/ui/async-await/in-trait/async-associated-types2-desugared.rs:
```diff
@@ -0,0 +1,31 @@
+    where
+        Self: 'a;
+
+    async fn foo(&self) -> Self::Fut<'a>;
```
Can you fix this? This should probably be `&'a self` (and adding `'a` generic lifetime) or `Self::Fut<'_>`.

nikomatsakis replied:
```suggestion
    fn foo(&self) -> Self::Fut<'a>;
```

I'm not 100% sure what this test is trying to test, but the name "desugared" suggests to me that it's trying to write out explicitly the equivalent of an `async fn`

bryangarza replied:
> Can you fix this? This should probably be `&'a self` (and adding `'a` generic lifetime) or `Self::Fut<'_>`.

oh, yeah that makes sense. I'll update it

bryangarza replied:
@spastorino what did you have in mind for this one (example #10 in your HackMD)?I know what desugared means in a general sense, not sure what that would look like for this though.

bryangarza replied:
Actually, I understand now, after re-reading the static AFIT RFC

compiler-errors commented on src/test/ui/async-await/in-trait/async-associated-types2.rs:
```diff
@@ -0,0 +1,32 @@
+    type Fut<'a> = impl Future + 'a
+    where
+        Self: 'a;
+    //~^^^ ERROR `impl Trait` in type aliases is unstable
```
You should enable this `type_alias_impl_trait` feature

compiler-errors commented on src/test/ui/async-await/in-trait/async-associated-types2.rs:
```diff
@@ -0,0 +1,32 @@
+    where
+        Self: 'a;
+
+    fn foo(&self) -> Self::Fut<'a>;
```
Same as above, this isn't the error we're looking for here -- can you fix the lifetime?

compiler-errors commented on src/test/ui/async-await/in-trait/async-generics-and-bounds.rs:
```diff
@@ -0,0 +1,21 @@
+// edition: 2021
```
This should have a bug filed for it

bryangarza replied:
Filed https://github.com/rust-lang/rust/issues/102682

compiler-errors commented on src/test/ui/async-await/in-trait/async-generics.rs:
```diff
@@ -0,0 +1,18 @@
+// edition: 2021
```
This can be linked to the same bug as above I think

compiler-errors commented on src/test/ui/async-await/in-trait/async-lifetimes-and-bounds.rs:
```diff
@@ -0,0 +1,20 @@
+// edition: 2021
```
This can be linked to the same bug as above I think

compiler-errors commented on src/test/ui/async-await/in-trait/fn-not-async-err2.rs:
```diff
@@ -0,0 +1,21 @@
+// edition: 2021
```
Can you add a variant of this test where `return_position_impl_trait_in_trait` is added? It probably will pass.

bryangarza replied:
will try it out

bryangarza replied:
Added :)

compiler-errors reviewed (commented):
Some initial comments, probably not comprehensive but wanted to do a first pass. I will take another look at it in a few days.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3178761476/jobs/5180580781) [(plain)](https://github.com/rust-lang/rust/commit/cc644cc42e9242db078c03f85900c790ed107774/checks/8690323845/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
* highest error code: E0790
Found 506 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/src/test/ui/async-await/in-trait/fn-not-async-err2.rs:11: trailing whitespace
some tidy checks failed
Build completed unsuccessfully in 0:00:10
```

</details>

nikomatsakis commented on src/test/ui/async-await/in-trait/async-associated-types.rs:
```diff
@@ -0,0 +1,24 @@
+    type MyAssoc = (&'a U, &'b T);
+
+    async fn foo(&'a self, key: &'b T) -> (&'a U, &'b T) {
+        (self, key)
```
I expect this test to pass -- what am I missing?

nikomatsakis replied:
(It may be that this is a known bug)

bryangarza replied:
Yep, filed https://github.com/rust-lang/rust/issues/102681

nikomatsakis commented on src/test/ui/async-await/in-trait/async-associated-types.rs:
```diff
@@ -0,0 +1,24 @@
+use std::fmt::Debug;
+
+trait MyTrait<'a, 'b, T> where Self: 'a, T: Debug + Sized + 'b {
+    type MyAssoc;// = (&'a T, &'b U);
```
```suggestion
    type MyAssoc;
```

What's this comment about?

bryangarza replied:
I think it was in the hackmd snippet, I'll just delete it

nikomatsakis reviewed (changes requested):
Gave a quick skim, though not over everything. Some questions!

bryangarza committed:
Update static AFIT tests based on feedback

compiler-errors commented on src/test/ui/async-await/in-trait/async-associated-types.rs:
```diff
@@ -0,0 +1,24 @@
+        (self, key)
+    }
+}
+//~^^^^ ERROR cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
```
nit: Can you change all of these to point at exactly the line they come from? There's no reason to do `^^^^` when you can move it up four lines and use `^`.

compiler-errors commented on src/test/ui/async-await/in-trait/async-generics-and-bounds.rs:
```diff
@@ -0,0 +1,21 @@
+// edition: 2021
```
You need to add `// known-bug: #0000000` to this.

bryangarza committed:
Update tests based on feedback

bryangarza committed:
Add additional tests for static AFIT

bryangarza commented:
Added some additional tests

compiler-errors commented on src/test/ui/async-await/in-trait/async-example-desugared-boxed.rs:
```diff
@@ -0,0 +1,23 @@
+
+impl MyTrait for i32 {
+    fn foo(&self) -> Pin<Box<dyn Future<Output = i32> + '_>> {
+        Box::pin(async {
```
This can't be marked with `// known-bug`, but this will be rejected by a PR that implements #102745, so perhaps say that this should/will be broken.

compiler-errors commented on src/test/ui/async-await/in-trait/async-example-desugared-in-trait.rs:
```diff
@@ -0,0 +1,20 @@
+}
+
+impl MyTrait for i32 {
+    async fn foo(&self) -> i32 {
```
Similarly, this can't be marked with `// known-bug`, but this will be rejected by a PR that implements #102745, so perhaps say that this should/will be broken.

compiler-errors commented on src/test/ui/async-await/in-trait/async-example-desugared.rs:
```diff
@@ -0,0 +1,22 @@
+}
+
+impl MyTrait for i32 {
+    fn foo(&self) -> impl Future<Output = i32> + '_ {
```
Similarly, this can't be marked with `// known-bug`, but this will be rejected by a PR that implements #102745, so perhaps say that this should/will be broken.

compiler-errors commented on src/test/ui/async-await/in-trait/async-example.rs:
```diff
@@ -0,0 +1,32 @@
+    let _ = x.bar();
+    // Calling from async block in non-async context
+    async {
+        let _ = x.foo();
+        let _ = x.bar();
```
```suggestion
        let _: i32 = x.foo().await;
        let _: i32 = x.bar().await;
```

makes the test a bit stronger

compiler-errors commented on src/test/ui/async-await/in-trait/async-lifetimes.rs:
```diff
@@ -0,0 +1,18 @@
+#![allow(incomplete_features)]
+
+trait MyTrait<'a, 'b, T> {
+    async fn foo(&'a self, key: &'b T) -> (&'a Self, &'b T);
```
This is basically identical to src/test/ui/async-await/in-trait/async-lifetimes-and-bounds.rs, so add the same known-bug.

compiler-errors commented on src/test/ui/async-await/in-trait/fn-not-async-err2-rpitit.rs:
```diff
@@ -0,0 +1,22 @@
+// check-pass
```
Is this identical to src/test/ui/async-await/in-trait/async-example-desugared.rs? If so, plz delete.

bryangarza replied:
oh yeah, it's the same. I'll get rid of it

compiler-errors reviewed (changes requested):
Left some more feedback.

bryangarza committed:
Update src/test/ui/async-await/in-trait/async-example.rs

Co-authored-by: Michael Goulet <michael@errs.io>

bryangarza committed:
Update tests based on feedback

- Add comment to some tests that will break when #102745 is implemented
- Mark a test with known-bug
- Delete duplicate test

bryangarza commented:
Thanks! I updated it now

compiler-errors commented:
This looks good for now. @bryangarza, make sure to note what has test coverage and what doesn't for future AFIT/RPITIT test writers. Not sure if there's a hackmd already for that.

@bors r+ rollup

bors commented:
:pushpin: Commit bfdefdbfb740a2b41dbdf01299ecdf1610c54907 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors bfdefdbfb740a2b41dbdf01299ecdf1610c54907 -->
<!-- homu: {"type":"Approved","sha":"bfdefdbfb740a2b41dbdf01299ecdf1610c54907","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
